python/core/auto_generated/qgsabstractdatabaseproviderconnection.sip.in

```diff
@@ -63,16 +63,49 @@ Returns the column names
 Returns the results rows
 %End
 
-        void appendColumn( const QString &columnName );
+        qlonglong rowCount() const;
 %Docstring
-Appends ``columnName`` to the list of column names.
+Returns the row count
+
+.. note::
+
+   the value may not be exact or it may be -1 if not known
 %End
 
-        void appendRow( const QList<QVariant> &row );
+        QList<QVariant> nextRow();
```

Comment:
I think it would be cleaner api wise to either go the "java style iterator" way with `nextRow()` and `hasNextRow()` to signal that the end is reached. Or an stl container approach with `[const]Begin` and `[const]End`.

Reply:
I thought about `hasNextRow()` but there are problems with some providers where we don't know the number of returned rows until we fetch them, and I've been using a forward only result, which is faster but has no seek functionality.

Is that really necessary? The current implementation returns an empty row to signal end of rows and it works just fine.

Reply:
Couldn't they just prefetch the next result?
It would be great if we could use language constructs like

```
QueryResult result = connection.execSql( sql );
for ( const Row &row : result )
{
  // do your magic
}
```

Reply:
I've done it with prefetch as you suggested.

Reply:
Excellent. Thanks for this nice implementation!